### PR TITLE
[feat][JAX] Add Apple Silicon MPS support

### PR DESCRIPTION
**Pull request by bvolpato**
## Summary

Add opt-in Apple Silicon support for SkyRL's JAX backend using `jax-mps`.

## Changes

- Add an `mps` dependency extra for Apple Silicon macOS.
- Restore dependency resolution for the current JAX and Transformers stack.
- Route causal attention through XLA when the jax-mps SDPA patch is active.
- Add MPS-compatible fallbacks for attention-mask and KV-cache updates.
- Preserve `dynamic_update_slice` boundary behavior in the KV-cache fallback.
- Keep ephemeral JAX sampler syncs in memory instead of writing temporary archives.
- Invalidate the in-memory sampler checkpoint ID after optimizer steps and training checkpoint restores so later snapshot loads cannot reuse stale weights.
- Document setup, supported workflows, and current limitations.
- Add focused backend, generator, attention, cache-boundary, and sampler-checkpoint coverage.

## Validation

The checkpoint-restore follow-up has two CPU regressions covering both optimizer-restore settings. Both fail before the fix and pass afterward using actual training and sampler checkpoint round trips. The focused optimizer and ephemeral-checkpoint controls also pass. Required pre-commit hooks pass.

Earlier Linux validation after rebasing onto `main` at `38ef74d3`:

- JAX backend, generator, and attention suites (`30 passed`)
- DeepSeek CPU suite (`9 passed` with eight forced XLA CPU devices)
- Sampler checkpoint and optimizer regressions (`2 passed`)
- Ruff, Black, and Gitleaks
- `uv lock --check` and `git diff --check`

The same branch code also passed the documentation type check and production build (`66 pages`), plus a focused type check with no branch-added diagnostics compared with `main`, before the final dependency-only rebase.

Earlier branch validation on an M4 Max covered the MPS JAX backend, DeepSeek-V3 against its Hugging Face reference, and Qwen3-0.6B forward, backward, optimizer update, and sampling. Those hardware results predate the latest rebase and fallback fixes; current fixes have CPU regression coverage but have not been rerun on Apple hardware.

## Limitations

- `jax-mps` remains experimental and supports one Apple GPU.
- Some JAX operations remain unsupported by the plugin.
- Causal attention uses the XLA implementation when the jax-mps patch is active.
- KV-cache updates use a dense `einsum` fallback until batched scatter support lands upstream.
- Buffer-donation warnings may still appear.

---

> [!NOTE]
> **Medium Risk**
> Changes shared JAX training/sampling weight lifecycle and attention/generation paths; bugs could cause wrong rollouts after optim steps, though new backend and unit tests target the main regressions.
> 
> **Overview**
> Adds **opt-in Apple Silicon** support for the JAX/Tinker backend via a new `mps` extra (`jax-mps`) and docs for running LoRA train/sample (and an optional RL smoke test) on macOS 14+.
> 
> **Runtime workarounds for jax-mps limits:** causal attention with padding masks routes through **XLA** when the MPS SDPA patch is active; decoding uses alternate **KV-cache** and **attention-mask** updates instead of `dynamic_update_slice` / `.at[...].set`.
> 
> **Training ↔ sampling correctness:** ephemeral `save_weights_for_sampler` can **skip disk** and mark in-memory weights as loaded; **optimizer steps** and **training checkpoint restores** clear `loaded_checkpoint_id` so sampling cannot reuse stale LoRA weights. Dependency overrides widen `transformers` / `ml_dtypes` resolution for macOS dev installs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 44292ffc2ba308b8dffcaf104775c9461502a454. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/tx/utils/generator.py [hunk 1]
```diff
@@ -85,6 +85,22 @@ def update_layer(kv_cache, k, v, positions):
         """
         k_cache, v_cache = kv_cache
 
+        if jax.default_backend() == "mps":
+            # jax-mps does not yet support the batched scatter produced by
+            # vmap(dynamic_update_slice).
+            max_start = max(k_cache.shape[1] - k.shape[1], 0)
+            update_start = positions[:, :1]
+            update_start = jnp.where(update_start < 0, update_start + k_cache.shape[1], update_start)
+            update_start = jnp.clip(update_start, 0, max_start)
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/tx/utils/generator.py:94]: This does not match JAX behavior. I checked `jax.lax.dynamic_update_slice(jnp.arange(5), jnp.array([90, 91]), (start,))` directly: starts `-7` and `-5` update `0:2`, `-4` updates `1:3`, `-1` updates `3:5`, and starts `4`, `5`, and `8` update `3:5`. JAX first wraps negative indices relative to operand size, then clamps to the valid update range. Current fallback preserves those results.

File: tests/tx/utils/test_generator.py [hunk 2]
```diff
@@ -77,6 +78,22 @@ def make_inputs(batch_size: int, prompt_length: int):
...
+    k_cache = jnp.zeros((3, 5, 1, 2), dtype=jnp.float32)
+    v_cache = jnp.zeros_like(k_cache)
+    k = jnp.arange(1, 13, dtype=jnp.float32).reshape(3, 2, 1, 2)
+    v = k + 10
+    positions = jnp.array([[-7, -6], [-4, -3], [4, 5]], dtype=jnp.int32)
+
+    updated_k, updated_v = KVCache.update_layer((k_cache, v_cache), k, v, positions)
+
+    expected_k = k_cache.at[0, 0:2].set(k[0]).at[1, 1:3].set(k[1]).at[2, 3:5].set(k[2])
+    expected_v = v_cache.at[0, 0:2].set(v[0]).at[1, 1:3].set(v[1]).at[2, 3:5].set(v[2])
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply tests/tx/utils/test_generator.py:92]: Current expectation is intentional and matches direct `dynamic_update_slice` output: with operand length 5 and update length 2, start `-4` resolves to `1:3`; start `-7` resolves to `0:2`; start `4` resolves to `3:5`. No test change needed.

File: skyrl/backends/jax.py [hunk 3]
```diff
@@ -986,17 +987,26 @@ def load_checkpoint(self, checkpoint_path: AnyPath, model_id: str, load_optimize
...
+                output_path,
+                self.process_id,
+            )
+            logger.info(f"Saved LoRA sampler checkpoint to {output_path}")
+
+        # Training and sampling share one in-memory model in this backend. Marking
+        # these weights as loaded avoids a redundant archive round trip on the
+        # next sample request, and lets ephemeral RL syncs skip disk entirely.
+        checkpoint_id = Path(str(output_path)).name.removesuffix(".tar.gz")
+        lora_model.loaded_checkpoint_id = checkpoint_id
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/jax.py:1012]: Thanks for catching the stale sampler ID. I found that training checkpoint restores need the same invalidation as optimizer steps, and fixed that in 8ef4c5a7. Both optimizer-restore settings now have a regression that saves one sampler snapshot, restores different training weights, and verifies that sampling reloads the requested snapshot. Both cases fail before the fix and pass afterward on CPU; Apple GPU validation is still outstanding.
